be/src/storage/binlog_file_reader.cpp

```diff
@@ -0,0 +1,348 @@
+        _current_file_pos += page_header_pb.compressed_size();
+        _next_page_index++;
+        // reach the end of the file
+        if (_next_page_index >= _file_meta->num_pages()) {
+            return Status::NotFound(
+                    strings::Substitute("Can't find version $0, changelog_id $1 "
+                                        "in file $2, largest version $3, seq_id $4, and changelog_id $5",
+                                        version, seq_id, _file_path, _file_meta->end_version(),
+                                        _file_meta->end_seq_id(), _file_meta->end_seq_id()));
+        }
+    } while (true);
+    _next_page_index++;
+    _init_current_log_entry();
+    return Status::OK();
+}
+
+Status BinlogFileReader::_read_next_page() {
+    if (_next_page_index >= _file_meta->num_pages()) {
+        return Status::EndOfFile("There is no more pages");
+    }
+
+    _reset_current_page_context();
+    PageHeaderPB& page_header_pb = _current_page_context->page_header;
+    RETURN_IF_ERROR(_read_page_header(_next_page_index, &page_header_pb));
+    RETURN_IF_ERROR(_read_page_content(_next_page_index, &page_header_pb, &(_current_page_context->page_content)));
+    _next_page_index++;
+    _init_current_log_entry();
+    return Status::OK();
+}
+
+Status BinlogFileReader::_seek_to_log_entry(int64_t seq_id) {
+    PageContentPB& page_content = _current_page_context->page_content;
+    int32_t num_log_entries = page_content.entries_size();
+    while (_current_page_context->next_log_entry_index < num_log_entries) {
+        _advance_log_entry();
+        int64_t end_seq_id = _current_log_entry->end_seq_id;
+        // end_seq_id == -1 is for an empty ingestion with only one EMPTY log entry
+        if (end_seq_id == -1 || end_seq_id >= seq_id) {
+            return Status::OK();
+        }
+    }
+
+    return Status::NotFound(strings::Substitute("Can't find log entry containing the change event <$0, $1>",
+                                                _current_page_context->page_header.version(), seq_id));
+}
+
+void BinlogFileReader::_advance_log_entry() {
+    PageContext* page_context = _current_page_context.get();
+    LogEntryInfo* log_entry_info = _current_log_entry.get();
+
+    int next_log_entry_index = page_context->next_log_entry_index;
+    LogEntryPB* next_log_entry = page_context->page_content.mutable_entries(next_log_entry_index);
+    int64_t next_start_seq_id = log_entry_info->end_seq_id + 1;
+    int num_change_events = 0;
+    // assume next log entry share file id with the current first
+    FileIdPB* next_file_id = log_entry_info->file_id;
+    // assume rows in next log entry are continuous with current log entry first
+    int next_start_row_id = log_entry_info->start_row_id + log_entry_info->num_rows;
+    int num_rows = 0;
+    switch (next_log_entry->entry_type()) {
+    case INSERT_RANGE_PB: {
+        InsertRangePB* entry_data = next_log_entry->mutable_insert_range_data();
```

Comment:
mutable_xx is used to write. I think we need to read. should call next_log_entry->insert_range_data() here?